iothub/device/src/Pipeline/RetryDelegatingHandler.cs

```diff
@@ -699,6 +663,55 @@ private async Task EnsureOpenedAsync(bool withRetry, CancellationToken cancellat
             }
         }
 
+        public override async Task CloseAsync(CancellationToken cancellationToken)
+        {
+            await _handlerSemaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+            try
+            {
+                if (!_openCalled)
+                {
+                    return;
+                }
+
+                if (Logging.IsEnabled)
+                    Logging.Enter(this, cancellationToken, nameof(CloseAsync));
+
+                _handleDisconnectCts.Cancel();
+                await base.CloseAsync(cancellationToken).ConfigureAwait(false);
+            }
+            finally
+            {
+                if (Logging.IsEnabled)
+                    Logging.Exit(this, cancellationToken, nameof(CloseAsync));
+
+                _handlerSemaphore?.Release();
+                Dispose(true);
+            }
+        }
+
+        private async Task EnsureOpenedAsync()
+        {
+            try
+            {
+                await _handlerSemaphore.WaitAsync().ConfigureAwait(false);
```

Comment:
Is a semaphore still required just to check connection state?

Reply:
few [unit tests](https://github.com/Azure/azure-iot-sdk-csharp/blob/95532a519f6081e6987b04b7f8a571e7e1e98671/iothub/device/tests/Pipeline/RetryDelegatingHandlerTests.cs#L21) were failing due to it. 
While digging into it, I found [this](https://github.com/Azure/azure-iot-sdk-csharp/blob/95532a519f6081e6987b04b7f8a571e7e1e98671/iothub/device/src/Pipeline/RetryDelegatingHandler.cs#L808) call which is invoked [here](https://github.com/Azure/azure-iot-sdk-csharp/blob/95532a519f6081e6987b04b7f8a571e7e1e98671/iothub/device/src/Pipeline/RetryDelegatingHandler.cs#L649)
Looks like it was due to semaphore behavior.

Reply:
So, yes?

Any way to find a way not to require it? Perhaps not in this PR, but in a future task?

Reply:
Yes, a semaphore is required in this case to check the connection state.

Removing the dependency sounds like a good task. I'll look more into it for a future change.